tools/openremoteslide-write-png.1
.\"
.\" OpenSlide, a library for reading whole slide image files
.\"
.\" Copyright (c) 2007-2012 Carnegie Mellon University
.\" All rights reserved.
.\"
.\" OpenSlide is free software: you can redistribute it and/or modify
.\" it under the terms of the GNU Lesser General Public License as
.\" published by the Free Software Foundation, version 2.1.
.\"
.\" OpenSlide is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
.\" GNU Lesser General Public License for more details.
.\"
.\" You should have received a copy of the GNU Lesser General Public
.\" License along with OpenSlide. If not, see
.\" <http://www.gnu.org/licenses/>.
.\"


.\" See man-pages(7) for formatting conventions.


.TH OPENSLIDE-WRITE-PNG 1 2012-02-24 "OpenSlide 341.0.1" "User Commands"

.mso www.tmac

.SH NAME
openremoteslide-write-png \- Write a region of a virtual slide to a PNG

.SH SYNOPSIS
.BR "openremoteslide-write-png " [ --help "] [" --version ]
.I slide-file x y level width height output-file

.SH DESCRIPTION
Write a region of the specified
.I level
of a virtual slide to a PNG image.
.I x
and
.I y
are pixel coordinates in level 0, the highest-resolution slide level.
.I width
and
.I height
are pixel dimensions in the specified
.IR level .

The dimensions of each level of a slide can be obtained with
.BR openremoteslide-show-properties (1).

.SH OPTIONS
.TP
.B --help
Display usage summary.

.TP
.B --version
Display version and copyright information.

.SH EXIT STATUS
.B openremoteslide-write-png
returns 0 on success, 1 if an error occurred, or 2 if the arguments are
invalid.

.SH COPYRIGHT
Copyright \(co 2007-2012 Carnegie Mellon University and others

OpenSlide is free software: you can redistribute it and/or modify it under
the terms of the
.URL http://gnu.org/licenses/lgpl-2.1.html "GNU Lesser General Public License, version 2.1" .

OpenSlide comes with NO WARRANTY, to the extent permitted by law.  See the
GNU Lesser General Public License for more details.

.SH SEE ALSO
.BR openremoteslide-quickhash1sum (1),
.BR openremoteslide-show-properties (1)


CHANGELOG.txt
Notable Changes in OpenSlide
============================

Version 341.0.1, 2015-04-20
 * New formats: Philips TIFF, Ventana TIFF
 * Support OpenJPEG 2.1.0
 * Improve performance of JPEG and JP2K decoding
 * Add openremoteslide.region[i].* properties
 * Improve MATLAB compatibility
 * Enable function deprecation warnings with MSVC
 * Many portability fixes
 * aperio: Detect OpenJPEG chroma subsampling breakage during open
 * aperio: Fill in missing tiles with downsampled data
 * aperio: Report MPP for slides scanned in locales with decimal comma
 * hamamatsu: Support NDPI files > 4 GB
 * hamamatsu: Properly detect NDPI slides produced by NDP.toolkit
 * hamamatsu: Support VMS/VMU slides without a NoLayers key
 * hamamatsu: Report MPP for VMS/VMU
 * leica: Support slides with 2010/03/10 XML namespace
 * leica: Base64-decode leica.barcode property in 2010/10/01 namespace
 * sakura: Support slides with multiple focal planes
 * sakura: Support slides without tile table
 * ventana: Support slides with multiple focal planes
 * ventana: Improve positioning of AOIs within level
 * ventana: Fix failure to recognize macro image on some slides

Version 3.4.0, 2014-01-25
 * Major internal restructuring
 * New formats: Hamamatsu NDPI, Sakura SVSLIDE, Ventana BIF (preliminary)
 * Add openremoteslide_detect_vendor()
 * Deprecate openremoteslide_can_open() (not very useful and often misused)
 * Document performance considerations for openremoteslide_open()
 * Add properties giving the bounds of the non-empty region of the slide
 * leica: Support multiple main images if their levels are coplanar
 * leica: Use slide size as level size
 * mirax: Support PNG- and BMP-formatted slides
 * mirax: Fix "Expected first 0 value" error opening some slides
 * mirax: Fix incorrect tile placement on some slides without overlaps
 * mirax: Never synthesize downsampled levels
 * Add OPENSLIDE_DEBUG environment variable (OPENSLIDE_DEBUG=? for help)
 * Fix some crashes in error paths
 * Add tests for many error paths

Version 3.3.3, 2013-04-13
 * Fix inclusion of openremoteslide.h with MSVC
 * Properly handle Aperio JP2K slides with zero-length tiles
 * Support Hamamatsu slides with blank MacroImage key

Version 3.3.2, 2012-12-01
 * Fix seams in MIRAX 2.2 slides (thanks, Agelos Pappas)
 * Fix associated image naming in single-level Aperio slides
 * Stop decoding MIRAX tiles outside requested region
 * Stop decoding unneeded tiles during tile-aligned accesses
 * Increase Hamamatsu VMU tile size to reduce rendering overhead
 * Document performance considerations for openremoteslide_can_open()

Version 3.3.1, 2012-10-14
 * Parallelize concurrent openremoteslide_read_region calls on an openremoteslide_t
 * Eliminate background scanning of tile headers in MIRAX
 * Scan many fewer tiles during first accesses to Hamamatsu VMS
 * Ignore Leica Z-planes other than 0
 * Add experimental tile-size properties
 * Document API thread safety

Version 3.3.0, 2012-09-08
 * Support Leica SCN format (requires libtiff >= 4) (thanks, Agelos Pappas)
 * Allow opening MIRAX 2.2 slides (though there are seams, bug #92)
 * Add standardized microns-per-pixel and objective-power properties
 * Add "macro" associated image in Trestle
 * Rename "layer" to "level" throughout the API (deprecate "layer" functions;
   remove "layer" properties)
 * Report parse errors in openremoteslide_open() by returning an openremoteslide_t in
   error state
 * Deprecate openremoteslide_get_comment()
 * Add openremoteslide_get_version()
 * Improve command-line tools; add manpages
 * Support building with MinGW-w64; drop CMake, MSVC, mingw32
 * Add tests for many error paths

Version 3.2.6, 2012-02-23
 * Support downsampled MIRAX files
 * Improve performance on MIRAX slides without tile overlaps
 * Fix openremoteslide_read_region for large dimensions on layer > 0
   (3.2.5 regression)
 * Correct subpixel error in MIRAX tile placement
 * Fix unlikely use-after-free with Hamamatsu VMU

Version 3.2.5, 2011-12-16
 * Support MIRAX 1.03 files (thanks, Jan Harkes)
 * Fix openremoteslide_read_region for large dimensions
 * Use subpixel precision in all backends
 * Don't keep associated images in memory
 * Disable quickhash-1 for TIFF files with very large top layer
 * Various build fixes (thanks, Jan, Marco Feuerstein, and
   Mathieu Malaterre)
 * Fix some unlikely memory leaks

Version 3.2.4, 2011-03-07
 * Support MIRAX files that do not have non-hierarchical data
   (thanks, Jan Harkes)
 * Fix compilation on Windows (thanks Hauke Heibel)
 * Work around a bug in GKeyFile parser (thanks, Jan)

Version 3.2.3, 2010-09-09
 * Support MIRAX files that use a variant format for tile
   positions (thanks, Hauke Heibel and Marco Feuerstein)
 * Update location of website
 * Add background color property, for slides that have it
 * Update CMake scripts and other Windows fixes
   (thanks Hauke and Marco)
 * Fix some test.c bugs
 * Fix incorrect MIRAX drawing at certain resolutions
   (thanks Hauke and Marco)
 * Support quickhash-1 on older systems (thanks, Jan Harkes)

Version 3.2.2, 2010-06-16
 * Rework some internals of openremoteslide_read_region
 * Support negative coordinates and zero-sized dimensions
   in openremoteslide_read_region
 * Clarify the documentation about openremoteslide_read_region
 * Fix Windows build bug with new NGR support
 * Enable untested BigTIFF support

Version 3.2.1, 2010-06-03
 * Fix crashes on Windows when trying to read Hamamatsu files
 * Fix jpeg 7 problems in read_from_one_jpeg
 * Quiet the error handling system after the first error

Version 3.2.0, 2010-06-01
 * Add experimental CMake support and fixes for building with MSVC
   (thanks to Hauke Heibel!)
 * Enable detecting runtime errors
 * Add initial Hamamatsu Nanozoomer VMU support
   (thanks to Steve Lamont)
 * Add "openremoteslide-write-png" tool

Version 3.1.1, 2010-04-27
 * Fix a crash when reading an invalid VMS file
 * Fix memory leaks when reading an invalid VMS file

tools/openremoteslide-write-png.c
/*
 *  OpenSlide, a library for reading whole slide image files
 *
 *  Copyright (c) 2007-2012 Carnegie Mellon University
 *  All rights reserved.
 *
 *  OpenSlide is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Lesser General Public License as
 *  published by the Free Software Foundation, version 2.1.
 *
 *  OpenSlide is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 *  GNU Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public
 *  License along with OpenSlide. If not, see
 *  <http://www.gnu.org/licenses/>.
 *
 */

#include "openremoteslide.h"
#include "openremoteslide-tools-common.h"
#include "openremoteslide-url.h"

#include <png.h>
#include <inttypes.h>
#include <glib.h>
#include <stdarg.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>
#include <setjmp.h>

static const char SOFTWARE[] = "Software";
static const char OPENREMOTESLIDE[] = "OpenSlide <http://openremoteslide.org/>";

#define ENSURE_NONNEG(i) \
  if (i < 0) {					\
    fail(#i " must be non-negative");	\
  }

#define ENSURE_POS(i) \
  if (i <= 0) {					\
    fail(#i " must be positive");	\
  }

static void fail(const char *format, ...) {
  va_list ap;

  va_start(ap, format);
  char *msg = g_strdup_vprintf(format, ap);
  va_end(ap);

  fprintf(stderr, "%s: %s\n", g_get_prgname(), msg);
  fflush(stderr);

  exit(1);
}


static void write_png(openremoteslide_t *osr, FILE *f,
		      int64_t x, int64_t y, int32_t level,
		      int32_t w, const int32_t h) {
  png_structp png_ptr = png_create_write_struct(PNG_LIBPNG_VER_STRING,
						NULL, NULL, NULL);
  if (!png_ptr) {
    fail("Could not initialize PNG");
  }

  png_infop info_ptr = png_create_info_struct(png_ptr);
  if (!info_ptr) {
    fail("Could not initialize PNG");
  }

  if (setjmp(png_jmpbuf(png_ptr))) {
    fail("Error writing PNG");
  }

  png_init_io(png_ptr, f);

  png_set_IHDR(png_ptr, info_ptr, w, h, 8,
	       PNG_COLOR_TYPE_RGB_ALPHA, PNG_INTERLACE_NONE,
	       PNG_COMPRESSION_TYPE_DEFAULT,
	       PNG_FILTER_TYPE_DEFAULT);

  // text
  png_text text_ptr[1];
  memset(text_ptr, 0, sizeof text_ptr);
  text_ptr[0].compression = PNG_TEXT_COMPRESSION_NONE;
  char *key = strdup(SOFTWARE);
  text_ptr[0].key = key;
  char *text = strdup(OPENREMOTESLIDE);
  text_ptr[0].text = text;
  text_ptr[0].text_length = strlen(text);

  png_set_text(png_ptr, info_ptr, text_ptr, 1);

  // background
  const char *bgcolor = openremoteslide_get_property_value(osr,
						     OPENREMOTESLIDE_PROPERTY_NAME_BACKGROUND_COLOR);
  if (bgcolor) {
    int r, g, b;
    sscanf(bgcolor, "%2x%2x%2x", &r, &g, &b);

    png_color_16 background = { 0, r, g, b, 0 };
    png_set_bKGD(png_ptr, info_ptr, &background);
  }

  // start writing
  png_write_info(png_ptr, info_ptr);

  uint32_t *dest = g_malloc(w * 4);
  int32_t lines_to_draw = h;
  double ds = openremoteslide_get_level_downsample(osr, level);
  int32_t yy = y / ds;
  while (lines_to_draw) {
    openremoteslide_read_region(osr, dest,
			  x, yy * ds, level, w, 1);

    const char *err = openremoteslide_get_error(osr);
    if (err) {
      fail("%s", err);
    }

    // un-premultiply alpha and pack into expected format
    for (int i = 0; i < w; i++) {
      uint32_t p = dest[i];
      uint8_t *p8 = (uint8_t *) (dest + i);

      uint8_t a = (p >> 24) & 0xFF;
      uint8_t r = (p >> 16) & 0xFF;
      uint8_t g = (p >> 8) & 0xFF;
      uint8_t b = p & 0xFF;

      switch (a) {
      case 0:
	r = 0;
	b = 0;
	g = 0;
	break;

      case 255:
	// no action
	break;

      default:
	r = (r * 255 + a / 2) / a;
	g = (g * 255 + a / 2) / a;
	b = (b * 255 + a / 2) / a;
	break;
      }

      // write back
      p8[0] = r;
      p8[1] = g;
      p8[2] = b;
      p8[3] = a;
    }

    png_write_row(png_ptr, (png_bytep) dest);
    yy++;
    lines_to_draw--;
  }

  // end
  g_free(dest);
  g_free(key);
  g_free(text);
  png_write_end(png_ptr, info_ptr);
  png_destroy_write_struct(&png_ptr, &info_ptr);
}


static const struct openremoteslide_tools_usage_info usage_info = {
  "slide x y level width height output.png",
  "Write a region of a virtual slide to a PNG.",
};

int main (int argc, char **argv) {
  _openremoteslide_tools_parse_commandline(&usage_info, &argc, &argv);
  if (argc != 8) {
    _openremoteslide_tools_usage(&usage_info);
  }

  // get args
  const char *slide = argv[1];
  int64_t x = g_ascii_strtoll(argv[2], NULL, 10);
  int64_t y = g_ascii_strtoll(argv[3], NULL, 10);
  int32_t level = strtol(argv[4], NULL, 10);
  int64_t width = g_ascii_strtoll(argv[5], NULL, 10);
  int64_t height = g_ascii_strtoll(argv[6], NULL, 10);
  const char *output = argv[7];

  // open slide
  openremoteslide_t *osr = openremoteslide_open(slide);

  // check errors
  if (osr == NULL) {
    fail("%s: Not a file that OpenSlide can recognize", slide);
  }

  const char *err = openremoteslide_get_error(osr);
  if (err) {
    fail("%s: %s", slide, err);
  }

  // validate args
  ENSURE_NONNEG(level);
  if (level > openremoteslide_get_level_count(osr) - 1) {
    fail("level %d out of range (level count %d)",
	 level, openremoteslide_get_level_count(osr));
  }
  ENSURE_POS(width);
  ENSURE_POS(height);
  if (width > INT32_MAX) {
    fail("width must be <= %d for PNG", INT32_MAX);
  }
  if (height > INT32_MAX) {
    fail("height must be <= %d for PNG", INT32_MAX);
  }

  // set up output file
  FILE *png = fopen(output, "wb");
  if (!png) {
    fail("Can't open %s for writing: %s", output,
	 strerror(errno));
  }

  write_png(osr, png, x, y, level, width, height);

  fclose(png);
  openremoteslide_close(osr);

  return 0;
}
